### ISSUE TITLE
pass async to end-user on start & end for sandbox

### ISSUE THREAD

**Issue by konsumer** [architect/architect#622]
This was discussed in slack.

**Is your feature request related to a problem? Please describe.**

Currently, `require('@architect/sandbox')` `end` and `start` functions swallow the callbacks , so end-user has no access to err or when the close has asynchronously finished. This is a sign that global async queue is being used, and breaks things like testing in jest (as it uses atomic async tests & before/after, not global queue.) It's not just a problem with testing though, it would break anything where you need to do something after `start` or `end`.

**Describe the solution you'd like**

If async were not swallowed, end-user could trigger their function when it's done.
I like promises, in general, and some will like callbacks, so this, as an example, supports both uses:

```js
const { promisify } = require('util')

function end(cb = () => {}) {
  return Promise.all(
    promisify(client.close)(),
    promisify(bus.close)(),
  ).then(() => {
    if ((arc.http || arc.ws) {
      return  promisify(http.close)().then(promisify(cb))
    }else{
      return promisify(cb)()
    }
  })
}
```
this resolves when `client` & `bus` `close` methods have been called in parallell, then `http` and user's callback. It returns a promise, and if the user gave a callback, it uses that (instead of requiring user to use a promise.)

**Comment by ryanblock** [comment]
I am working on attempting to reproduce this problem, but am running into other issues in Jest.

Currently, when starting Sandbox in Jest, I receive the following error:

```sh
/Users/ryan/git/jesty/node_modules/@architect/sandbox/src/db/_init.js:28
    dynamo.createTable({
           ^

TypeError: dynamo.createTable is not a function
    at Array._createBackUpSessions (/Users/ryan/git/jesty/node_modules/@architect/sandbox/src/db/_init.js:28:12)
    at runWaterfall (/Users/ryan/git/jesty/node_modules/run-waterfall/index.js:26:13)
    at init (/Users/ryan/git/jesty/node_modules/@architect/sandbox/src/db/_init.js:92:3)
    at Server._server (/Users/ryan/git/jesty/node_modules/@architect/sandbox/src/db/index.js:37:9)
    at Object.onceWrapper (events.js:312:28)
    at Server.emit (events.js:223:5)
    at emitListeningNT (net.js:1340:10)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
npm ERR! Test failed.  See above for more details.
```

Code relates back to this `require` call: https://github.com/architect/sandbox/blob/master/src/db/_init.js#L10-L12

When Sandbox is started by Jest, it appears is not loading the module, and is instead instantiating it as an empty object:
```js
  // Don't load dynamo client with global requires, it may not yet have env vars loaded
  // eslint-disable-next-line
  dynamo = require('./_get-db-client')
  console.log('dynamo:', dynamo)
```
```sh
  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "dynamo: {}".

      at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:186:10)
      at init (node_modules/@architect/sandbox/src/db/_init.js:14:11)
      at Server._server (node_modules/@architect/sandbox/src/db/index.js:37:9)
```

What I'd expect to see logged from a properly instantiated Dynamo client instance from that module:
```js
dynamo: Service {
  config: Config {
    credentials: SharedIniFileCredentials {
...
```

A reason I do not personally use Jest is because of the magical things it does under the hood in rewriting module resolvers, mutating globals, etc. Unfortunately, it appears that this is the kind of behavior I'm currently seeing in attempting to reproduce this issue. (For reference: this issue goes away if I require the module globally in `src/db/_init.js` ¯\\\_(ツ)\_\/¯)

Anywho, as I don't have much knowledge in sidestepping Jest's magic, any pointers on how we can get it to stop messing with Node's module resolution and/or whatever else may be going on under the hood that's interfering with Sandbox running deterministically?

**Comment by konsumer** [comment]
> Currently, when starting Sandbox in Jest, I receive the following error:

After a clean install & update of arc to no RC, I am getting this too.

> A reason I do not personally use Jest is because of the magical things it does under the hood in rewriting module resolvers, mutating globals, etc. 

I don't super-care if it's jest, I can use mocha/chai in my immediate demo thing. It's my goto nowadays as it has everything built in (test, expect, full mocking, etc. It's like mocha+chai+sinon+more.) I will say that generally when I see it not working with a lib, it's a sign the lib is doing the same kind of stuff, which for anything other than a testing/mocking library is bad practice. Hopefully mocha won't care, and it will work. I'll give it a try.

In this case, I think this comment points to the issue:
```
// Don't load dynamo client with global requires, it may not yet have env vars loaded
```
It's doing dynamic require & relying on side-effects of import, which in many situations can be very bad, not just jest. I work with react/webpack/parcel/pkg at work (pkg being the only one of those I could see immediately being useful with arc, like zeit-now uses it to make smaller lambdas) and it would break in all of these, among many others. I think even babel might have an issue with it, and typescript would too. I can help clean it up, which would probly make jest work. I think the very simplest solution would be to export a function there that does same side-effects, and call it at that point, instead of requiring it. I think my original feature-request still  applies, though, like if the user had access to the async mechanism (via promise or callback) this wouldn't be such a big issue, because you could do fixes after it was imported.

Related:

* #621  - If we had at-function option-resolution (JS default params are resolved at runtime, so you can get around the original issue pretty elegantly.)

**Comment by ryanblock** [comment]
I think we just crossed paths here and on Slack! I think I've managed to resolve all issues in `1.5.9-RC`, please `npm i @architect/sandbox@RC` and let me know how it goes.

Per the Dynamo thing, worked around that by putting that operation into a callback fn, seems fine and more deterministic, so that's a nice bonus. 📈

**Comment by konsumer** [comment]
> Per the Dynamo client thing, I worked around that by putting that operation into a callback fn, seems fine and more deterministic, so that's a nice bonus. chart_with_upwards_trend

Nice, we came to the same conclusion. I will test, now.

**Comment by konsumer** [comment]
[Here](https://github.com/konsumer/arc-issue-622) is a minimal test-case. Looks good!

**Comment by konsumer** [comment]
I spoke too soon. Looks like `require('@architect/functions')` has some other async side-effects.

If I add this to top of test file:

```js
const arc = require('@architect/functions')
```

I get the same error: `TypeError: end is not a function`:

```
 FAIL  ./jest.test.js
  ● Test suite failed to run

    TypeError: end is not a function

      17 | afterAll(async () => {
      18 |   console.log('Stopping sandbox')
    > 19 |   await end()
         |         ^
      20 | })
      21 | 
      22 | describe('jest', () => {

      at Object.<anonymous> (jest.test.js:19:9)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        6.473s
Ran all test suites.
  console.log jest.test.js:18
    Stopping sandbox

Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

**Comment by ryanblock** [comment]
Diff module's side effects, let's scope that bug to that module. This fix will ship in `1.6.0`!
